src/plugins/ccode/ccode.cpp

```diff
@@ -0,0 +1,438 @@
+ */
+void decodeKey (Key * key, CCodeData * mapping)
+{
+	const char * value = static_cast<const char *> (keyValue (key));
+	if (!value) return;
+
+	keySetRaw (key, mapping->buffer, decode (value, keyGetValueSize (key), mapping));
+}
+
+/**
+ * @brief This function replaces unescaped characters in a key value with escaped characters.
+ *
+ * The function stores the escaped result value both in `mapping->buffer` and the given key.
+ *
+ * @pre The variable `mapping->buffer` needs to be twice as large as the key value’s size.
+ *
+ * @param key This key stores the value this function escapes.
+ * @param mapping This variable stores the buffer and the character mapping this function uses to encode the value of the given key.
+ */
+void encodeKey (Key * key, CCodeData * mapping)
+{
+	const char * value = static_cast<const char *> (keyValue (key));
+	if (!value) return;
+
+	keySetRaw (key, mapping->buffer, encode (value, keyGetValueSize (key), mapping));
+}
+
+// ====================
+// = Plugin Interface =
+// ====================
+
+/** @see elektraDocOpen */
+int elektraCcodeOpen (Plugin * handle, Key * key ELEKTRA_UNUSED)
+{
+	CCodeData * const mapping = new CCodeData ();
+
+	/* Store for later use...*/
+	elektraPluginSetData (handle, mapping);
+
+	KeySet * const config = elektraPluginGetConfig (handle);
+
+	Key const * const escape = ksLookupByName (config, "/escape", 0);
+	mapping->escape = '\\';
+	if (escape && keyGetBaseNameSize (escape) && keyGetValueSize (escape) == 3)
+	{
+		int escapeChar = elektraHexcodeConvFromHex (keyString (escape)[1]);
+		escapeChar += elektraHexcodeConvFromHex (keyString (escape)[0]) * 16;
+
+		mapping->escape = escapeChar & 255;
+	}
+	ELEKTRA_LOG_DEBUG ("Use “%c” as escape character", mapping->escape);
+
+	Key const * const root = ksLookupByName (config, "/chars", 0);
+
+	if (root)
+	{
+		readConfig (mapping, config, root);
+	}
+	else
+	{
+		setDefaultConfig (mapping);
+	}
+
+	return ELEKTRA_PLUGIN_STATUS_NO_UPDATE;
+}
+
+/** @see elektraDocClose */
+int elektraCcodeClose (Plugin * handle, Key * key ELEKTRA_UNUSED)
+{
+	CCodeData * const mapping = static_cast<CCodeData *> (elektraPluginGetData (handle));
+
+	delete[](mapping->buffer);
+	delete (mapping);
+
+	return ELEKTRA_PLUGIN_STATUS_NO_UPDATE;
+}
+
+/** @see elektraDocGet */
+int elektraCcodeGet (Plugin * handle, KeySet * returned, Key * parentKey)
+{
+	if (!strcmp (keyName (parentKey), "system/elektra/modules/ccode"))
+	{
+		KeySet * const pluginConfig = contract ();
+		ksAppend (returned, pluginConfig);
+		ksDel (pluginConfig);
+		return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	}
+
+	CCodeData * const mapping = retrieveMapping (handle);
+
+	CppKeySet keys{ returned };
+	CppKeySet unescaped{};
+	for (auto key : keys)
+	{
+		size_t const size = max (key.getBinarySize (), key.getNameSize ());
+		if (size > mapping->bufferSize)
+		{
+			mapping->bufferSize = size;
+			mapping->buffer = new unsigned char[mapping->bufferSize];
```

Comment:
In modern C++ "new" is avoided. Here there is a big memory leak. You should use an array/vector instead. It avoids the memory leak, size is then implicitely available, and vectors have resize().

Reply:
> In modern C++ "new" is avoided.

I know. I just replaced `malloc` with `new` and `free` with `delete`.

> Here there is a big memory leak.

I do not think so, since I **did not change the way the plugin handles memory**. If there is a memory leak, then the Linux Travis ASAN build should also fail, which [is not the case](https://travis-ci.org/ElektraInitiative/libelektra/builds/390917873).

> You should use an array/vector instead. It avoids the memory leak, size is then implicitely available, and vectors have resize().

This certainly makes sense (except for the part about the memory leak of course 😊).

Reply:
The memory leak only occurs if the size is > 1000 (especially in gradually increasing sizes there might be a memleak in every iteration), seems like this is not tested. I did not say you introduced the problem but in C++ it is now easier to fix :smile: 

And in general it is not safe to translate every malloc -> new because C++ additionally has exceptions which can cause memleaks not possible in C.